### ISSUE TITLE
Add  `__str__` implementations for DataTypes with a more compact version of `__repr__`

### ISSUE THREAD

**Issue by stinodego** [pola-rs/polars#13439]
Inspired by https://github.com/pola-rs/polars/pull/13357

Currently, printing a data type can result in quite long strings, for example, when an `Enum` has many categories. Datetimes also get quite wordy.

I propose we implement `__str__` with a more compact representation. We omit argument names, and omit parentheses for types without arguments.

The `__repr__` version gets the fully qualified string representation of the data type, which are guaranteed to result in the data type when `eval`'d. We include parentheses for instantiated types.

Examples:

| Type | `str` | `repr` |
| --- | --- | --- |
| `Int8` | `Int8` | `Int8` |
| `Int8()` | `Int8` | `Int8()` |
| `Datetime` | `Datetime` | `Datetime` |
| `Datetime()` | `Datetime('us')` | `Datetime(time_unit='us', time_zone=None)` |
| `Datetime("us", "UTC")` | `Datetime('us', 'UTC')` | `Datetime(time_unit='us', time_zone='UTC')` |
| `Enum("abcdefg")` | `Enum(['a', 'b', 'c' ... 'e', 'f', 'g'])` | `Enum(categories=['a', 'b', 'c', 'd', 'e', 'f', 'g'])` |

I think this can lead to a better experience during quick prototyping, as printed data types are easier to read. And we retain the current functionality for users who want the fully qualified string.

Thoughts?

**Comment by mcrumiller** [comment]
I'm a fan of this. What about using brackets instead of parens for details? e.g. `Datetime['us']` versus `Datetime('us')`? This is how numpy/pandas show dtype details, and IMO makes it look less like a function and more of a property of the dtype.

**Comment by stinodego** [comment]
That's definitely worth considering!

**Comment by ritchie46** [comment]
Isn't that a bad idea if we have 1 million categories? Shouldn't we just make the datatypes support pickle and json?

**Comment by mcrumiller** [comment]
@ritchie46 yeah I think that case maybe. FYI [here](https://docs.python.org/3/library/functions.html#repr) is the text from the official python docs (emphasis mine):

> Return a string containing a printable representation of an object. For many types, this function makes an attempt to return a string that would yield an object with the same value when passed to [eval()](https://docs.python.org/3/library/functions.html#eval); **otherwise, the representation is a string enclosed in angle brackets that contains the name of the type of the object together with additional information often including the name and address of the object**.

We could set a threshold, or we could just say "if the user really wanted it, here it is" and give them a super long string.

**Comment by ritchie46** [comment]
>if the user really wanted it, here it is" and give them a super long string.

Yes, but I don't expect calling `repr` to block my thread if we have a large `Enum` type. 

For serialization in understand and expect that this costs compute.

**Comment by stinodego** [comment]
> Isn't that a bad idea if we have 1 million categories?

Probably... we can set an upper bound. It can be the same as the `__str__` or maybe a bit higher.

> Shouldn't we just make the datatypes support pickle and json?

We can definitely add a `json_encode` / `json_decode` to the `DataType` class. That doesn't have to block this improvement. I'm not sure what a JSON representation of a DataType would look like, though. I guess something like:

```json
{
  "Enum": {
    "categories": ["a", "b"]
  }
}
```

**Comment by alexander-beedie** [comment]
Hmm... Think I'm with @ritchie46 on this one; it's not the case that anyone would _expect_ that the `repr(obj)` can be run through `eval` to recreate it, and there's also no reason _why_ the str version would be more manageable than the repr - don't forget that the fully-expanded repr result is what would be shown for every `frame.schema`[^1] repr result  too...

We don't do this for DataFrame, LazyFrame, or Expr, so I'm not sure why we'd decide to do it specially for DataType given the potential downsides (I can't see that there's a particularly compelling reason to do so; serialisation should be handled with dedicated methods, and that was probably the main driver for wanting it).

Note that the phrasing in the Python docs is:

_"For **many** types..."_ ('many', not 'all')
_"...this function **makes an attempt**"_ ('attempt')

In other words, this approach is considered _optional_, and cannot be relied on anywhere else either; we are under no obligation to do so ;)

[^1]: Looking at the frame schema is likely the most common way that the DataType repr will be viewed, and it wouldn't be the proposed short/str one.

**Comment by stinodego** [comment]
Hm, for some reason I thought the `__str__` version was used in the Python console / Notebooks, but I was wrong. That definitely limits the usefulness of this whole venture.

**Comment by mcrumiller** [comment]
@stinodego in the console, if create an object `x` and simply type the `x` and hit enter in the command prompt, it gives you the result of `__repr__`. If you type `print(x)` it gives you the result of `__str__`:

```shell
>>> class TestReprStr:
...     def __str__(self):
...         return "__str__"
...     def __repr__(self):
...         return "__repr__"
... 
>>> x = TestReprStr()
>>> x
__repr__
>>> print(x)
__str__
```

**Comment by mcrumiller** [comment]
I decided to add some `__str__` implementations just to see what a more compact version might look like:

```python
import polars as pl

dt_u8 = pl.UInt8
dt_u16 = pl.UInt16
dt_u32 = pl.UInt32
dt_u64 = pl.UInt64
dt_i8 = pl.Int8
dt_i16 = pl.Int16
dt_i32 = pl.Int32
dt_i64 = pl.Int64
dt_dec = pl.Decimal(precision=5, scale=2)
dt_dur = pl.Duration("ms")
dt_time = pl.Time
dt_date = pl.Date
dt_datetime = pl.Datetime("ns")
dt_datetime_tz = pl.Datetime("ns", "UTC")
dt_array = pl.Array(inner=pl.UInt8, width=3)
dt_list = pl.List(inner=pl.Datetime("ms"))
dt_struct = pl.Struct(fields={"a": pl.Int8, "b": dt_list, "c": pl.Int32, "d": dt_array})

print(f"""
{dt_u8}\n{dt_u16}\n{dt_u32}\n{dt_u64}\n{dt_i8}\n{dt_i16}
{dt_i32}\n{dt_i64}\n{dt_dec}\n{dt_dur}\n{dt_time}\n{dt_date}
{dt_datetime}\n{dt_datetime_tz}\n{dt_array}\n{dt_list}\n{dt_struct}
""")
```
```
UInt8
UInt16
UInt32
UInt64
Int8
Int16
Int32
Int64
Decimal[5,2]
Duration[ms]
Time
Date
Datetime[ns]
Datetime[ns,UTC]
Array[UInt8,3]
List[Datetime[ms]]
Struct[a[Int8];b[List[Datetime[ms]]];c[Int32];d[Array[UInt8,3]]]
```

The struct looks a bit messy but I'm not sure that's avoidable.

**Comment by stinodego** [comment]
Struct could be `Struct[a:Int8,c:Int32]` etc. in your syntax.

But yeah, I am giving this one a re-think as the benefits aren't as obvious as I thought. And having `print` show something else than on the console can also be a little confusing.

**Comment by alexander-beedie** [comment]
I don't think we should rely on non-obvious differences between `__str__` and `__repr__` without a good reason; if we want a compact repr (which certainly isn't a bad idea) we should make that a discoverable method, rather than a "magic" behaviour of `__str__` that is DataType specific 🤔 

I'd suggest the following:

* Default repr for large Enum objects is truncated, as per https://github.com/pola-rs/polars/pull/13357. This ensures that frame schemas (and dtypes) will always be reasonable by default (in notebooks, console, logs, etc).
* Add dedicated serialisation methods (to/from JSON would be consistent with other objects; pickle should "just work").
* Add a dedicated compact repr method for places where the caller wants to opt for something minimal.
